### DISCUSSION TITLE
Element relations

### DISCUSSION THREAD

**Discussion by Thorras** [IfcOpenShell/IfcOpenShell#2656] [General]
Hello fellow nerd.
I am fairly new to python and ifcopenshell.
And i wonder, is it possible to get to an element based on what sone it relates to?
Like a bathroom connected to a wall and then a hallway on the other side.
As long as the geometry is cut accordingly i should work, right?

And if it does, perhaps someone could point me in the direction of an answer?

Thank you :)

Thor

**Comment by aothms** [comment]
You could either use `ifcopenshell.geom.tree` and then `.select(extend=0.01)` (or some other small number) to get geometrical neighbours. This requires evaluating all the geometry which is potentially computationally expensive. It also requires some manual filtering of the results because you would get also vertical neighbours which you do not seem interested in. The other alternative is to depend on space boundaries (IfcRelSpaceBoundary) provided by the authoring tool to encode the information of which element bound which space. This can in some cases be less reliable as authoring tools cannot always do this reliably (the geometrical proximity is also easier to compute then defining the geometric bounded surface of the interface) but since this depends on information provided at the authoring stage it's more efficient to compute.

**Comment by Thorras** [comment]
Thanks for a quick response! I'll give it a go :)

**Comment by DprOliv** [comment]
@aothms  ... let's suppose that we need to get all the doors providing access to a space ... and also that the authoring tool encoded the boundaries elements of spaces (walls mostly). 
Which path should i use to get the walls (and then, inside the walls, the doors ...) ? Would the use of model.get_inverse(space) help me getting the walls defining the boundary of the desired space ? Or should i use model.traverse(space) ?
And then, once we get the walls defining the boundary. Would the usage of ifcRelVoidsElement() then ifcOpeningElement() then ifcRelFillsElement () give me those doors ?

Could you send a segment of Python code exemplifying such approach ?

**Reply by aothms** [reply]
> model.get_inverse(space) or should i use model.traverse(space) ?

You don't have to use either. The relationship from Space to RelSpaceBoundary is directly available as a named inverse attributes `BoundedBy` http://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSpace.htm#5.4.3.64.3-Attributes

The second part is correct

Wall - RelVoids - Opening - RelFills - Door/Window

I don't have any code for this at hand. If you post some attempts here, I'm sure the community can further assist.

**Reply by DprOliv** [reply]
Thanks for the quick reply Thomas.

I tried to play with BoundedBy attribute ... as you suggested ...  
I used the sequence below to traverse each space and then checked each boundary element, and for each i can then verify it it's a wall and proceed with the search for doors in them ...

for space in list_spaces:
   for boundData in space.BoundedBy:
         boundInfo = boundData.get_info()
         bound_element = boundInfo["RelatedBuildingElement"]
         info_element = bound_element.get_info()
         print(info_element)

is this approach the one you recommended ?

**Reply by aothms** [reply]
Yes.

Minor think: you do not have to use `boundData.get_info()["RelatedBuildingElement"]`. `boundData.RelatedBuildingElement` is the same and more efficient because it doesn't need to build a dictionary for all attribute values.

**Reply by DprOliv** [reply]
Got it... thanks for the guidance Thomas.
